### PR TITLE
Improve explanation in rescore

### LINKED ISSUE #28725: Why is score different from value in explanation?
Hi,

I did an explain request and get the following response:

```
                "_score": 2,
                "fields": {
                    "title": [
                        "Title"
                    ],
                    "id": [
                        "vbce77e7t9pe"
                    ]
                },
                "highlight": {
                    "content.search": [
                        "... some content ..."
                    ]
                },
                "_explanation": {
                    "value": 196.23717,
                    "description": "sum of:",
                    "details": [
                        {...
                        },
                        {
                            "value": 4.737492,
                            "description": "product of:",
                            "details": [...
                            ]
                        }
                    ]
                },
```
As you can see the score value is 2 and the value in the explanation is 196.23717.

Why is this different?

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Currently in a rescore request if window_size is smaller than
the top N documents returned (N=size), explanation of scores could be incorrect
for documents that were a part of topN and not part of rescoring.
This PR corrects this by saving in RescoreContext docIDs of documents
for which rescoring was applied, and adding rescoring explanation
only for these docIDs.

Closes #28725

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@elasticmachine run gradle build tests

File: server/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java [hunk 1]
```diff
@@ -92,7 +97,7 @@ public Explanation explain(int topLevelDocId, IndexSearcher searcher, RescoreCon
 
         // NOTE: we don't use Lucene's Rescorer.explain because we want to insert our own description with which ScoreMode was used.  Maybe
         // we should add QueryRescorer.explainCombine to Lucene?
-        if (rescoreExplain != null && rescoreExplain.isMatch()) {
+        if (rescoreContext.isRescored(topLevelDocId) && rescoreExplain != null && rescoreExplain.isMatch()) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java]: We always build the rescore explanation even when `rescoreContext.isRescored(topLevelDocId)` is false. We could avoid this by building the primary explanation first and return it directly if `rescoreContext.isRescored(topLevelDocId)` is false ?

**Review by jimczi — COMMENTED** [review]
Thanks @mayya-sharipova 
I left a small comment regarding the creation of the rescore explanation.

**Review by jimczi — APPROVED** [review]
LGTM

**Review by jpountz — APPROVED** [review]

**Pull request merged**